src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WinRTWebSocket.cs

```diff
@@ -279,47 +279,60 @@ public override void Dispose()
-                    if (args.MessageType == SocketMessageType.Binary)
-                    {
-                        messageType = WebSocketMessageType.Binary;
-                    }
-                    else
+                    uint dataAvailable;
+                    while ((dataAvailable = reader.UnconsumedBufferLength) > 0)
                     {
-                        messageType = WebSocketMessageType.Text;
-                    }
+                        ArraySegment<byte> buffer;
+                        try
+                        {
+                            buffer = _receiveAsyncBufferTcs.Task.GetAwaiter().GetResult();
+                        }
+                        catch (OperationCanceledException) // Caused by Abort call on WebSocket
+                        {
+                            return;
+                        }
 
-                    bool endOfMessage = false;
-                    uint readCount = Math.Min(dataAvailable, (uint) buffer.Count);
-                    var dataBuffer = reader.ReadBuffer(readCount);
-                    // Safe to cast readCount to int as the maximum value that readCount can be is buffer.Count.
-                    dataBuffer.CopyTo(0, buffer.Array, buffer.Offset, (int) readCount);
-                    if (dataAvailable == readCount)
-                    {
-                        endOfMessage = true;
-                    }
+                        _receiveAsyncBufferTcs = new TaskCompletionSource<ArraySegment<byte>>();
+                        WebSocketMessageType messageType;
+                        if (args.MessageType == SocketMessageType.Binary)
+                        {
+                            messageType = WebSocketMessageType.Binary;
+                        }
+                        else
+                        {
+                            messageType = WebSocketMessageType.Text;
+                        }
 
-                    WebSocketReceiveResult recvResult = new WebSocketReceiveResult((int) readCount, messageType,
-                        endOfMessage);
-                    _webSocketReceiveResultTcs.TrySetResult(recvResult);
+                        bool endOfMessage = false;
+                        uint readCount = Math.Min(dataAvailable, (uint) buffer.Count);
+                        var dataBuffer = reader.ReadBuffer(readCount);
+                        // Safe to cast readCount to int as the maximum value that readCount can be is buffer.Count.
+                        dataBuffer.CopyTo(0, buffer.Array, buffer.Offset, (int) readCount);
+                        if (dataAvailable == readCount)
+                        {
+                            endOfMessage = true;
+                        }
+
+                        WebSocketReceiveResult recvResult = new WebSocketReceiveResult((int) readCount, messageType,
+                            endOfMessage);
+                        _webSocketReceiveResultTcs.TrySetResult(recvResult);
```

Comment:
I do believe that we want the continuation to run on this thread, considering that:

- ClientWebSocket supports at most one ReceiveAsync operation at any given time.
- WinRT invokes MessageReceived events in serial fashion, waiting for our handler to finish executing before raising the next event.

If we were to use the `TaskCreationOptions.RunContinuationsAsynchronously` option, we may end up processing MessageReceived events at a higher rate than what the .NET app is able to handle via ReceiveAsync, which would lead to increased memory footprint (internally-buffered data would be ever-increasing in size).